### PR TITLE
Calculate closest target position based on robot pose

### PR DESCRIPTION
**Pull request by james-ward**
Tested on robot 16 Apr 23

### CODE REVIEW

File: utilities/game.py [hunk 1]
```diff
@@ -19,3 +20,25 @@ def is_red() -> bool:
...
+    return Translation2d(pose3d.x, pose3d.y)
+
+
+def find_closest_tag(robot_pose: Pose2d) -> tuple[Pose3d, int]:
+    tag_ids = [1, 2, 3] if is_red() else [6, 7, 8]
+
+    # Use the first tag to set a baseline for distance
+    best_id = tag_ids[0]
+    best_tag = apriltag_layout.getTagPose(best_id)
+    closest_distance = robot_pose.translation().distance(pose3d_to_translation2d(best_tag))  # type: ignore
```
Comment (auscompgeek) [inline utilities/game.py:35]: I presume the type check failure is because of the optional?
```suggestion
    assert best_tag is not None
    closest_distance = robot_pose.translation().distance(pose3d_to_translation2d(best_tag))
```
Alternatively we could get all of the tags in this module and assert they're all not None, like we're already doing for a couple of them.

File: utilities/game.py [hunk 2]
```diff
@@ -19,3 +20,25 @@ def is_red() -> bool:
...
+
+    for tag_id in tag_ids[1:]:
+        tag = apriltag_layout.getTagPose(tag_id)
+        d = robot_pose.translation().distance(pose3d_to_translation2d(tag))  # type: ignore
```
Comment (auscompgeek) [inline utilities/game.py:39]:
```suggestion
        assert tag is not None
        d = robot_pose.translation().distance(pose3d_to_translation2d(tag))
```

File: utilities/game.py [hunk 3]
```diff
@@ -19,3 +20,25 @@ def is_red() -> bool:
...
+        if d < closest_distance:
+            closest_distance = d
+            best_tag = tag
+            best_id = tag_id
+    return best_tag, best_id  # type: ignore
```
Comment (auscompgeek) [inline utilities/game.py:44]:
```suggestion
    return best_tag, best_id
```

Comment (LucienMorey) [inline utilities/game.py:44]: I think it might be better to assert that the best tag is none, so the type checker works. This part of the function is static. It should never break without someone changing it, and its something I think we want to know about instead of ignoring it.

I also thought I would highlight that there is an existing api to do what you are trying to do with the transform to 2d but if you just prefer what you have then that's cool too.

TLDR - I think we should assert that the tag is not none because its a static thing and should never break under reasonable conditions but idm if you prefer to keep your existing to 2d function. I will approve after this is done or I am happy to change it myself

```suggestion
def find_closest_tag(robot_pose: Pose2d) -> tuple[Pose3d, int]:
    tag_ids = [1, 2, 3] if is_red() else [6, 7, 8]

    # Use the first tag to set a baseline for distance
    best_id = tag_ids[0]
    best_tag = apriltag_layout.getTagPose(best_id)
    
    assertbest_tag != None
    
    closest_distance = robot_pose.translation().distance(
        best_tag.toPose2d().translation()
    )

    for tag_id in tag_ids[1:]:
        tag = apriltag_layout.getTagPose(tag_id)
        d = robot_pose.translation().distance(
        tag.toPose2d().translation()
    )
        if d < closest_distance:
            closest_distance = d
            best_tag = tag
            best_id = tag_id
    return best_tag, best_id
```

**Review by LucienMorey — CHANGES_REQUESTED** [review]
I think this looks good. I notice the

File: utilities/game.py (hunk 1 above)
Comment (LucienMorey) [reply utilities/game.py:35]: lmao we both basically did this at the same time

**Comment by james-ward** [comment]
Happy to use the built-in API to transform to 2D.

But running a bunch of useless `assert`s on stuff we *know* cannot be invalid (we set which tag IDs we look for about 10 lines above) just to shut up a linter is not cool, IMO.

**Comment by auscompgeek** [comment]
> But running a bunch of useless asserts on stuff we know cannot be invalid (we set which tag IDs we look for about 10 lines above) just to shut up a linter is not cool, IMO.

`assert` is disabled when running robot code on the roboRIO, so they won't have any effect on the real robot, if that's what you're concerned about.

**Comment by james-ward** [comment]
We have now managed to go from code that passed all the checks when I originally pushed it up, to code that is broken.

**Comment by auscompgeek** [comment]
> I suggest we just tell the linter to take a chill pill for a line or two.

Alright, how about #33 then?

**Comment by james-ward** [comment]
> > I suggest we just tell the linter to take a chill pill for a line or two.
> 
> Alright, how about #33 then?

That's a nice solution. We should use it, but I have fixed this PR now. We should switch later. But thanks for the solution, I prefer it.

**Comment by LucienMorey** [comment]
I had a go but no dice. My method meant that things passed locally but still threw their toys out of the cot up here. I will keep thinking about this during the day

**Comment by auscompgeek** [comment]
Jeez. Not even using `typing.Final` with tuples narrows to the correct literal types in mypy (when pyright does):
```py
tag_ids: typing.Final = (1, 2, 3) if is_red() else (4, 5, 6)
```

**Review by LucienMorey — APPROVED** [review]
